kythe/typescript/indexer.ts

```diff
@@ -965,6 +993,60 @@ class Visitor {
     return undefined;
   }
 
+  getSymbolAndVNameForFunctionDeclaration(node: ts.FunctionLikeDeclaration):
+      [ts.Symbol|undefined, VName|undefined] {
+    let context: Context|undefined = undefined;
+    if (ts.isGetAccessor(node)) {
+      context = Context.Getter;
+    } else if (ts.isSetAccessor(node)) {
+      context = Context.Setter;
+    }
+    if (node.name) {
+      const sym = this.host.getSymbolAtLocation(node.name);
+      if (!sym) { return [undefined, undefined]; }
+      const vname =
+          this.host.getSymbolName(sym, TSNamespace.VALUE, context);
+      return [sym, vname];
+    } else {
+      // TODO: choose VName for anonymous functions and return symbol
+      return [undefined, this.newVName('TODO', 'TODOPath')];
+    }
+  }
+
+  /**
+   * Given a node finds a function node that contains given node and returns it.
+   * If the node is not inside a function - returns SourceFile node. Examples:
+   *
+   * function foo() {
+   *   var b = 123;
+   * }
+   * var c = 567;
+   *
+   * For 'b' node this function will return 'foo' node.
+   * For 'c' node this function will return SourceFile node.
+   */
+  getContainingFunctionNode(node: ts.Node): ts.FunctionLikeDeclaration|ts.SourceFile {
+    node = node.parent;
+    for (; node.kind !== ts.SyntaxKind.SourceFile; node = node.parent) {
+      const kind = node.kind;
+      if (kind === ts.SyntaxKind.FunctionDeclaration ||
+          kind === ts.SyntaxKind.ArrowFunction ||
+          kind === ts.SyntaxKind.MethodDeclaration ||
+          kind === ts.SyntaxKind.Constructor ||
+          kind === ts.SyntaxKind.GetAccessor ||
+          kind === ts.SyntaxKind.SetAccessor ||
+          kind === ts.SyntaxKind.MethodSignature) {
+        return node as ts.FunctionLikeDeclaration;
+      }
+    }
+    return node as ts.SourceFile;
+  }
+
+  getSyntheticFileInitVName(): VName {
+    return this.newVName(
+        'fileInit:synthetic', this.host.moduleName(this.file.fileName));
```

Comment:
What is a fileInit:synthetic?  Is this a Kythe thing?

Reply:
It's arfiticial function node. It's needed so that each `ref/call` anchor can be a parent of a function even if if `ref/call` anchor is called on top-level. Name doesn't matter, I just used it as I think it can't conflict with any other TS name.